File: Linked List/Intersection of Two Linked Lists/getIntersectionNode.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
 /*
 在算法思路上参考了这篇博客http://blog.sina.com.cn/s/blog_725dd1010100tqwp.html
 */
struct ListNode *getIntersectionNode(struct ListNode *headA, struct ListNode *headB) {
    if(!headA || !headB) return NULL;
    struct ListNode *p = headA;
    while(p->next)
        p = p->next;
    p->next = headB;//将B结点接到A结点后面
    struct ListNode *fast=headA,*slow=headA;
    struct ListNode *mark;
    while(fast&& fast->next)
    {
        fast = fast->next->next;
        slow = slow->next;
        if(fast == slow)//两者相遇，说明存在环，记录相遇的位置为mark
        {
            mark = fast;
            break;
        }
    }
    if(!fast || !fast->next) //没有形成环，说明两者之间没有交集
    {
        p->next = NULL;
        return NULL;
    }
    else
    {
        struct ListNode *start = headA;
        while(start != mark) //start从头结点开始，mark从环中相遇的位置开始，两者相遇的地方就是环的入口
        {
            start = start->next;
            mark = mark->next;
        }
        p->next = NULL;
        return mark;  
    }
}

File: Linked List/Intersection of Two Linked Lists/getIntersectionNode.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        if headA is None or headB is None:
            return None
        p = headA
        while p.next:
            p = p.next
        p.next = headB
        fast = slow = headA
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
            if fast==slow:
                mark = fast
                break
        if not fast or not fast.next:
            p.next = None
            return None
        else:
            start = headA
            while start != mark:
                start = start.next
                mark = mark.next
            p.next = None
            return mark
